vortex-expr/src/pack.rs

```diff
@@ -0,0 +1,233 @@
+use std::any::Any;
+use std::fmt::Display;
+use std::sync::Arc;
+
+use itertools::Itertools as _;
+use vortex_array::aliases::hash_set::HashSet;
+use vortex_array::array::StructArray;
+use vortex_array::validity::Validity;
+use vortex_array::{ArrayData, IntoArrayData};
+use vortex_dtype::field::Field;
+use vortex_dtype::FieldNames;
+use vortex_error::{vortex_bail, VortexResult};
+
+use crate::{ExprRef, VortexExpr};
+
+#[derive(Debug, Clone)]
+pub struct Pack {
+    names: FieldNames,
+    values: Vec<ExprRef>,
+    validity: Option<ExprRef>,
+}
+
+impl Pack {
+    pub fn try_new_expr(
+        names: FieldNames,
+        values: Vec<ExprRef>,
+        validity: Option<ExprRef>,
+    ) -> VortexResult<Arc<Self>> {
+        if names.len() != values.len() {
+            vortex_bail!("length mismatch {} {}", names.len(), values.len());
+        }
+        if names.len() < 1 {
+            vortex_bail!("must provide at least one field");
```

Comment:
Why?

Reply:
Empty struct, but you will need a length there, maybe have a new ctor

Reply:
Too late on a Friday. I was worried about not knowing the length, but our expressions are evaluated on exactly one array and cannot change the length, so we know the resulting length.

Reply:
Done.